python/graphstorm/eval/utils.py

```diff
@@ -213,6 +214,172 @@ def calc_dot_neg_head_score(heads, tails, num_chunks, chunk_size,
+
+        Score function of RotateE measures the angular distance between
+        head and tail elements. The angular distance is defined as:
+
+        .. math:
+
+            d_r(h, t)=\|h\circ r-t\|
+
+        The RotatE score function is defined as:
+
+        .. math:
+
+            gamma - \|h\circ r-t\|^2
+
+        where gamma is a margin.
+
+        For more detials please refer to https://arxiv.org/abs/1902.10197
+        or https://dglke.dgl.ai/doc/kg.html#rotatee.
+
+        Parameters
+        ----------
+        h_emb: th.Tensor
+            Head node embedding.
+        t_emb: th.Tensor
+            Tail node embedding.
+        r_emb: th.Tensor
+            Relation type embedding.
+        rel_emb_init: float
+            The initial value used to bound the relation embedding initialization.
+        gamma: float
+            The gamma value used for shifting the optimization target.
+
+        Return
+        ------
+        rotate_score: th.Tensor
+            The RotatE score.
+    """
+    if device is not None:
+        r_emb = r_emb.to(device)
+        h_emb = h_emb.to(device)
+        t_emb = t_emb.to(device)
+
+    re_head, im_head = th.chunk(h_emb, 2, dim=-1)
+    re_tail, im_tail = th.chunk(t_emb, 2, dim=-1)
+
+    phase_rel = r_emb / (rel_emb_init / th.tensor(math.pi))
+    re_rel, im_rel = th.cos(phase_rel), th.sin(phase_rel)
+    re_score = re_head * re_rel - im_head * im_rel
+    im_score = re_head * im_rel + im_head * re_rel
+    re_score = re_score - re_tail
+    im_score = im_score - im_tail
+    score = th.stack([re_score, im_score], dim=0)
+    score = score.norm(dim=0)
+
+    rotate_score = gamma - score.sum(-1)
+    return rotate_score
+
+def calc_rotate_neg_head_score(heads, tails, r_emb, num_chunks,
+                               chunk_size, neg_sample_size,
+                               rel_emb_init, gamma,
+                               device=None):
+    """ Calculate RotatE Score for negative pairs when head nodes are negative.
+
+        Parameters
+        ----------
+        h_emb: th.Tensor
+            Head node embedding.
+        t_emb: th.Tensor
+            Tail node embedding.
+        r_emb: th.Tensor
+            Relation type embedding.
+        num_chunks: int
+            Number of shared negative chunks.
+        chunk_size: int
+            Chunk size.
+        neg_sample_size: int
+            Number of negative samples for each positive node.
+        rel_emb_init: float
+            The initial value used to bound the relation embedding initialization.
+        gamma: float
+            The gamma value used for shifting the optimization target.
+
+        Return
+        ------
+        rotate_score: th.Tensor
+            The RotatE score.
+    """
+    if device is not None:
+        r_emb = r_emb.to(device)
+        heads = heads.to(device)
+        tails = tails.to(device)
+    hidden_dim = heads.shape[1]
+    emb_real = tails[..., :hidden_dim // 2]
```

Comment:
Why using `hidden_dim // 2` here?

Reply:
The node embedding is composed of real part and imaginary part.